### PR TITLE
+ inner join with translation table

### PR DESCRIPTION
**Pull request by approached**
Inner join with the translation table.

Usage:
```php
$nodes = ProductCategory::JoinTranslation()->get();
```

Creates just one fast query.

### CODE REVIEW

**Comment by dimsav** [comment]
Hi @approached!

Unfortunately I still couldn't find the time to review your PR. Will do my best to get it done this week. Thanks for your patience :)

**Comment by sdebacker** [comment]
That's a really good feature when working with many items. I want it since https://github.com/dimsav/laravel-translatable/issues/139.

File: src/Translatable/Translatable.php [hunk 1]
```diff
@@ -607,4 +612,42 @@ protected function locale()
...
+    /**
+     * Inner join with the translation table
+     *
+     * @param \Illuminate\Database\Eloquent\Builder $query
+     * @param null $locale
+     * @return this
+     */
+    public static function scopeJoinTranslation(Builder $query, $locale = null)
+    {
+        // models
```
Comment (dimsav) [inline src/Translatable/Translatable.php]: Please remove this comment

File: src/Translatable/Translatable.php [hunk 2]
```diff
@@ -607,4 +612,42 @@ protected function locale()
...
+        $instance = new static;
+
+        $translationModelName = $instance->getTranslationModelName();
+        $translationModel = new $translationModelName();
+
+        // locale
```
Comment (dimsav) [inline src/Translatable/Translatable.php]: Remove this comment as well.

File: src/Translatable/Translatable.php [hunk 3]
```diff
@@ -57,6 +58,10 @@ public function getTranslation($locale = null, $withFallback = null)
         $locale = $locale ?: $this->locale();
         $withFallback = $withFallback === null ? $this->useFallback() : $withFallback;
         $fallbackLocale = $this->getFallbackLocale($locale);
+        
+        if($this->isJoinTranslated()){
```
Comment (dimsav) [inline src/Translatable/Translatable.php]: Please put a space between the `if` and the `(`. Also a space before the final `{`

Like this: `if ($this->isJoinTranslated()) {`.

File: tests/TranslatableTest.php [hunk 4]
```diff
@@ -411,4 +411,17 @@ public function test_fallback_for_country_based_locales()
...
         $this->assertSame('french fries', $fries->getTranslation('en-US')->name);
     }
+
+    public function test_it_returns_the_translation_with_inner_join()
+    {
+        DB::enableQueryLog();
+        $this->app->setLocale('el');
+
+        /** @var Country $country */
+        $country = Country::JoinTranslation()->first();
```
Comment (dimsav) [inline tests/TranslatableTest.php]: small caps `j` like this: `$country = Country::joinTranslation()->first();`

File: src/Translatable/Translatable.php [hunk 5]
```diff
@@ -607,4 +612,42 @@ protected function locale()
...
+        return $query->join($translationModel->getTable() . ' as t', 't.' . $instance->getRelationKey(), '=', $instance->getTable() . '.' . $instance->getKeyName())
+            ->where($instance->getLocaleKey(), $locale);
+    }
+    
+    /**
+     * Check if translation was with a inner join
+     *
+     * @return bool
+     */
+    protected function isJoinTranslated(){
```
Comment (dimsav) [inline src/Translatable/Translatable.php]: Make this a private method.

File: src/Translatable/Translatable.php [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
 trait Translatable
 {
+
```
Comment (dimsav) [inline src/Translatable/Translatable.php]: Remove this new line

**Comment by dimsav** [comment]
Very nice improvement! Please  fix those minor issues and the conflicts to merge this :)

**Comment by dimsav** [comment]
Also we need to update the docs before we release this.

**Comment by webfactorybulgaria** [comment]
+1

File: src/Translatable/Translatable.php [hunk 7]
```diff
@@ -607,4 +612,42 @@ protected function locale()
...
+        $translationModelName = $instance->getTranslationModelName();
+        $translationModel = new $translationModelName();
+
+        // locale
+        if (is_null($locale)) {
+            $locale = $instance->locale();
+        }
+
+        return $query->join($translationModel->getTable() . ' as t', 't.' . $instance->getRelationKey(), '=', $instance->getTable() . '.' . $instance->getKeyName())
```
Comment (webfactorybulgaria) [inline src/Translatable/Translatable.php]: Maybe left join would be better as a translation may not yet exist.

File: src/Translatable/Translatable.php [hunk 8]
```diff
@@ -607,4 +612,42 @@ protected function locale()
...
+    }
+    
+    /**
+     * Check if translation was with a inner join
+     *
+     * @return bool
+     */
+    protected function isJoinTranslated(){
+        foreach($this->translatedAttributes as $translatedAttribute){
+            if(isset($this->original[$translatedAttribute]) && !empty($this->original[$translatedAttribute])){
```
Comment (webfactorybulgaria) [inline src/Translatable/Translatable.php]: There is a case when $this->original[$translatedAttribute] is empty string - this doesn't mean it is not translated. In these cases an additional query is issued to the database.

File: src/Translatable/Translatable.php (hunk 7 above)
Comment (dimsav) [inline src/Translatable/Translatable.php]: We need tests for that.

File: src/Translatable/Translatable.php (hunk 8 above)
Comment (dimsav) [inline src/Translatable/Translatable.php]: Good point. Let's add some tests here as well.

**Comment by approached** [comment]
Thanks for checking the code. I'm not sure when i have enough time to fix this code.

**Comment by approached** [comment]
What status is here? Can you merge it?

File: tests/TestsBase.php [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
-use Orchestra\Testbench\TestCase;
 use Dimsav\Translatable\Test\Model\Country;
+use Orchestra\Testbench\TestCase;
 
 class TestsBase extends TestCase
 {
     protected $queriesCount;
 
-    const DB_NAME     = 'translatable_test';
+    const DB_NAME = 'translatable_test';
```
Comment (dimsav) [inline tests/TestsBase.php]: please revert this

File: tests/TestsBase.php [hunk 10]
```diff
@@ -23,19 +23,19 @@ public function setUp()
     }
 
     /**
-     * return void
+     * return void.
```
Comment (dimsav) [inline tests/TestsBase.php]: Please revert

File: tests/TestsBase.php [hunk 11]
```diff
@@ -48,10 +48,10 @@ private function runQuery($query)
         $dbPassword = static::DB_PASSWORD;
 
         $command = "mysql -u $dbUsername ";
-        $command.= $dbPassword ? " -p$dbPassword" : "";
-        $command.= " -e '$query'";
+        $command .= $dbPassword ? " -p$dbPassword" : '';
+        $command .= " -e '$query'";
 
-        exec($command . " 2>/dev/null");
+        exec($command.' 2>/dev/null');
```
Comment (dimsav) [inline tests/TestsBase.php]: revert.

File: tests/TestsBase.php [hunk 12]
```diff
@@ -93,7 +93,7 @@ protected function countQueries()
         $event->listen('illuminate.query', function ($query, $bindings) use ($that) {
             $that->queriesCount++;
             $bindings = $this->formatBindingsForSqlInjection($bindings);
-            $query    = $this->insertBindingsIntoQuery($query, $bindings);
+            $query = $this->insertBindingsIntoQuery($query, $bindings);
```
Comment (dimsav) [inline tests/TestsBase.php]: revert

**Comment by dimsav** [comment]
I really want to merge this but you guys are not helping. By keeping PRs small and to the point we get less conflicts, comments and it's easier to merge.

**Comment by approached** [comment]
Thanks for your time.
I will use my fork package https://github.com/approached/laravel-translatable

**Comment by nWidart** [comment]
Too bad this didn't make into this package. 

@approached You forgot this is an MIT licensed package, and removed the credits to @dimsav in the `LICENSE` file.

**Comment by approached** [comment]
@nWidart Thank you for the info. I renamed the files.

File: tests/TestsBase.php (hunk 9 above)
Comment (k-zakhariy) [reply tests/TestsBase.php]: what's the strange formatting  ? 😄 seems changes are ok

**Comment by k-zakhariy** [comment]
hey @approached , could you share your code ?

**Comment by approached** [comment]
https://gist.github.com/approached/2c9eab2b247d0540d14a90dc061a98ad

method: `scopeJoinTranslation` and `scopeWhereTranslationLike`

I prefer to create a **trait** in your project and use it in your _models_.

And by the way, the package maintainer is shitty. 👎

**Comment by k-zakhariy** [comment]
Big thanks @approached

**Comment by dimsav** [comment]
> And by the way, the package maintainer is shitty. 👎

@approached I apologize for making you feel this way.

**Pull request closed without merging**